### PR TITLE
improve client tx result handling

### LINKED ISSUE #20: refactor `TxResult` event
We only need to have the `"applied.hash"` field separately for the tendermint query when we're waiting for tx result in the client. For everything else, we should serialize the whole `TxResult` into a single field, so we can use the same type in the client to decode it. 

Currently we're doing this manually in `impl From<serde_json::Value> for TxResponse {` in `apps/src/lib/client/tx.rs` and in `fn finalize_block` in `apps/src/lib/node/ledger/shell.rs`

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1201757896515792/1201762912053900) by [Unito](https://www.unito.io)

### PR DESCRIPTION
**Pull request by tzemanovic**
## Describe your changes

- closes #20

The `TxResponse` type is updated to contain a typed `inner_tx: TxResult` for inner txs and some of its other fields are now typed. This is to make it possible to check a tx result in the client before taking further actions that depend on the result.

Stdout from a transaction now looks like this:
```
Transaction added to mempool.
Wrapper transaction hash: 137DE5305AC348B216D956AC538E6D23E233B20847C7B7EC266F48F2C3BA2CE1
Inner transaction hash: 89A8DBE501CC141508FBD54B9ED747AE3800A3BA299045323A7139988BBA3472
Wrapper transaction accepted at height 202. Used 21 gas.
Waiting for inner transaction result...
Transaction was successfully applied at height 203. Used 7890 gas.
```

A failed tx prints more info, e.g.:
```
Transaction added to mempool.
Wrapper transaction hash: 6E9083C5786E144B2BB190467FDC392129E8BC8D59AA1B394A4606F4E1EC5E02
Inner transaction hash: 619B2531D37A5F7F81787B4D5401158579647DDBEFB18BD9C5D4C3414B429777
Wrapper transaction accepted at height 172. Used 20 gas.
Waiting for inner transaction result...
Transaction was rejected by VPs: [
  "tnam1q9rhgyv3ydq0zu3whnftvllqnvhvhm270qxay5tn"
].
Changed keys: [
  "#tnam1pyqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqej6juv/#tnam1q9mjvqd45u7w54kee2aquugtv7vn7h3xrcrau7xy/balance/#tnam1q9rhgyv3ydq0zu3whnftvllqnvhvhm270qxay5tn",
  "#tnam1pyqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqej6juv/#tnam1q9mjvqd45u7w54kee2aquugtv7vn7h3xrcrau7xy/balance/#tnam1qxfj3sf6a0meahdu9t6znp05g8zx4dkjtgyn9gfu"
]
```

## Indicate on which release or other PRs this topic is based on

#2218 - diff range for review https://github.com/anoma/namada/pull/2276/files/dd28da14cefd6806ebfe71eb6a8ba8c894d975cd..f167f2b5b8396acf148fcc9cf992daf89f3b48a7

## Checklist before merging to `draft`
- [x] I have added a changelog
- [x] Git history is in acceptable state

### CODE REVIEW

File: apps/src/lib/client/tx.rs [hunk 1]
```diff
@@ -951,23 +953,23 @@ pub async fn submit_transfer(
...
+                    // And it is rejected by a VP
+                    matches!(resp.inner_tx_result(), InnerTxResult::VpsRejected(_)) =>
                 {
-                    // Then we probably straddled an epoch boundary. Let's retry...
-                    edisplay_line!(namada.io(),
-                        "MASP transaction rejected and this may be due to the \
-                        epoch changing. Attempting to resubmit transaction.",
-                    );
-                    continue;
+                    let submission_epoch = rpc::query_and_print_epoch(namada).await;
```
Comment (tzemanovic) [inline apps/src/lib/client/tx.rs:963]: this prevents querying epoch when it's not needed

File: core/src/types/transaction/mod.rs [hunk 2]
```diff
@@ -33,6 +36,119 @@ use crate::types::ibc::IbcEvent;
...
+    Copy,
+    Clone,
+    FromPrimitive,
+    ToPrimitive,
+    PartialEq,
+    Eq,
+    Serialize,
+    Deserialize,
+)]
+pub enum ResultCode {
```
Comment (sug0) [inline core/src/types/transaction/mod.rs:52]: nit: maybe document that these codes shouldn't be changed around between versions, so people can rely on their integer values to identify errors

File: sdk/src/tx.rs [hunk 3]
```diff
@@ -128,17 +129,24 @@ pub enum ProcessTxResponse {
...
-    Dump,
+    DryRun(TxResult),
 }
 
 impl ProcessTxResponse {
-    /// Get the the accounts that were reported to be initialized
-    pub fn initialized_accounts(&self) -> Vec<Address> {
+    // Is the transaction applied and was it accepted by all VPs? Note that this
+    // always returns false for dry-run transactions.
+    pub fn is_applied_and_valid(&self) -> bool {
```
Comment (sug0) [inline sdk/src/tx.rs:138]: oh we do have a method to check both

File: sdk/src/tx.rs [hunk 4]
```diff
@@ -315,23 +327,18 @@ pub async fn broadcast_tx(
     )?;
 
     if response.code == 0.into() {
-        display_line!(
-            context.io(),
-            "Transaction added to mempool: {:?}",
-            response
-        );
+        display_line!(context.io(), "Transaction added to mempool.");
+        tracing::debug!("Transaction mempool response: {:?}", response);
```
Comment (sug0) [inline sdk/src/tx.rs:331]: should we use the verbose debug print?
```suggestion
        tracing::debug!("Transaction mempool response: {response:#?}");
```

File: tests/src/e2e/ibc_tests.rs [hunk 5]
```diff
@@ -1264,35 +1265,28 @@ fn transfer(
...
-        .rsplit_once(' ')
+        .split_once('.')
         .unwrap()
-        .1
-        .replace(['"', ','], "");
-    if code != "0" {
-        return Err(eyre!("The IBC transaction failed: unread {}", unread));
-    }
+        .0;
+    dbg!(&height_str);
```
Comment (sug0) [inline tests/src/e2e/ibc_tests.rs:1288]:
```suggestion
```

File: tests/src/integration/masp.rs [hunk 6]
```diff
@@ -1014,7 +1014,7 @@ fn masp_txs_and_queries() -> Result<()> {
                 "--node",
                 validator_one_rpc,
             ],
-            Response::Ok("Transaction is valid"),
+            Response::Ok("Transaction was successfully applied"),
```
Comment (sug0) [inline tests/src/integration/masp.rs:1017]:
```suggestion
            Response::Ok(TX_APPLIED_SUCCESS),
```
maybe we should make this a const somewhere? next time we change this, if any, will result in a shorter diff

**Review by sug0 — CHANGES_REQUESTED** [review]
couple of suggestions

File: apps/src/lib/client/tx.rs [hunk 7]
```diff
@@ -951,23 +953,23 @@ pub async fn submit_transfer(
...
+                    // And its submission epoch doesn't match construction epoch
+                    if tx_epoch.unwrap() != submission_epoch {
+                        // Then we probably straddled an epoch boundary. Let's retry...
+                        edisplay_line!(namada.io(),
+                            "MASP transaction rejected and this may be due to the \
+                            epoch changing. Attempting to resubmit transaction.",
+                        );
+                        continue;
```
[the remark this answers is not in the document]
Comment (Fraccaman) [reply apps/src/lib/client/tx.rs:971]: this will be remove with one of marco's PRs

File: tests/src/integration/masp.rs (hunk 6 above)
Comment (tzemanovic) [reply tests/src/integration/masp.rs:1017]: good call, will add it to e2e tests too

Comment (tzemanovic) [reply tests/src/integration/masp.rs]: I've added bunch of strings in https://github.com/anoma/namada/pull/2276/files/f167f2b5b8396acf148fcc9cf992daf89f3b48a7..d09bc55133bfb8dddbadb9b9f5aa6ae2739718d6

File: core/src/types/transaction/mod.rs (hunk 2 above)
Comment (tzemanovic) [reply core/src/types/transaction/mod.rs:53]: done in [c0fc84b](https://github.com/anoma/namada/pull/2276/commits/c0fc84bd272cb0b4904d4440444018f543ab36c7)

File: apps/src/lib/client/tx.rs [hunk 8]
```diff
@@ -328,9 +328,11 @@ where
...
         signing::generate_test_vector(namada, &tx).await?;
 
-        let result = namada.submit(tx, &args.tx).await?;
-        if let ProcessTxResponse::Applied(response) = result {
-            return Ok(response.initialized_accounts.first().cloned());
+        let response = namada.submit(tx, &args.tx).await?;
+        if let ProcessTxResponse::Applied(tx_resp) = response {
+            if let InnerTxResult::Success(result) = tx_resp.inner_tx_result() {
+                return Ok(result.initialized_accounts.first().cloned());
+            }
```
[the remark this answers is not in the document]
Comment (tzemanovic) [reply apps/src/lib/client/tx.rs]: done in [42f05af](https://github.com/anoma/namada/pull/2276/commits/42f05af34bba7edcdc196e4e63bdccb36c497609)

**Pull request merged**